File: trees/balanced-binary-tree.cpp
// Given a binary tree, determine if it is height-balanced

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int isBalancedHelper(TreeNode* root, bool& isBalanced) {
        if (!root) return 0;
        int left = isBalancedHelper(root->left, isBalanced);
        int right = isBalancedHelper(root->right, isBalanced);

        if (abs(left - right) > 1) {
            isBalanced = false;
            return -1;
        }

        return 1 + max(left, right);

    }

public:
    bool isBalanced(TreeNode* root) {
        bool balanced = true;
        isBalancedHelper(root, balanced);
        return balanced;
    }
};

File: trees/balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        balancedFlag = True


        def checkHeight(root: Optional[TreeNode]) -> int:
            nonlocal balancedFlag
            
            if not root:
                return 0
            
            left = checkHeight(root.left)
            right = checkHeight(root.right)

            if (abs(left - right) > 1):
                balancedFlag = False
                return -1

            return 1 + max(left, right)


        checkHeight(root)
        return False if not balancedFlag else True
